README
Speech::Recognizer::SPX - Perl bindings for Sphinx-II
-----------------------------------------------------
David Huggins-Daines <dhuggins@cs.cmu.edu>
http://www.cs.cmu.edu/~dhuggins
June 3, 2005

Installation
------------

This module requires Sphinx2 0.5 or later, available from
http://www.sourceforge.net/projects/cmusphinx/.  To build and install
Sphinx-II, unpack the sphinx2 tarball, change into the directory
created and run:

  ./configure
  make
  # now become root, or some user with write access to /usr/local
  make install

On GNU/Linux systems, you may need to add /usr/local/lib to the end of
your /etc/ld.so.conf file and re-run /sbin/ldconfig as root.

Once you have installed Sphinx2, you can build and install
Speech::Recognizer::SPX.  To do this, change into the
Speech-Recognizer-SPX directory, and run:

  perl Makefile.PL
  make
  make test # this may fail if you have no working audio device
  make install

If any failures are reported by 'make test' that are not
audio-related, please report them by e-mail to the address above.

This module has only been tested on GNU/Linux.  It is likely to work
on free 4.4BSD-derived operating systems as well, but I have not
tested this.  Status on other Unix-like operating systems supported by
Sphinx-II, or Microsoft Windows, is unknown.

Demos and examples
------------------

Several example programs are included in this distribution:

tty-continuous: Basically the same sa the 'sphinx2-continuous' program
	included with the Sphinx-II distribution.

turtle-demo: Basically the same as the 'sphinx2-demo' program included
	with the Sphinx-II distribution.

server-demo: Demo of using a separate process to do recognition.

fe-test: Demo of the feature extraction model.

select-cad: Demo of using select() on audio devices in conjunction
	with Audio::SPX::Continuous.


Changes
Revision history for Perl extension Speech::Recognizer::SPX

0.08_01 Wed Feb 22 23:21:22 EST 2006
	- Compatibility fixes for post-0.6 CVS versions of Sphinx2

0.08 Fri Aug  5 16:35:37 EDT 2005
	- Official release with N-best, segmentation.

0.07_92 Mon Jul 18 10:30:18 EDT 2005
	- Incorporate proper backwards-compatibility code.

0.07_91 Tue Jul 12 12:31:51 EDT 2005
     	- Fix compilation/test on platforms without audio
	- Try to be backwards-compatible with Sphinx2 0.3 (untested)
	- Add support for word segmentation in uttproc
	- Add support for N-best lists, funded by SingleTouch
          Interactive, Inc (http://www.singletouch.net/)

0.07 Fri Jun  3 12:23:46 EDT 2005
	- Long-overdue update
	- Restores compatibility with Sphinx2 0.5
	- Makes it possible to use a non-standard --sphinx-prefix
	- New contact addresses, etc.

0.06_02 Tue Mar 27 12:17:45 EST 2001
	- First externally released version.
	- No code changes from 0.06_01, just different copyright
	  notices, with the addition of a README.

0.06_01 Fri Feb  9 12:52:49 EST 2001 (UNRELEASED)
	- Give a way to pass raw data into Audio::SPX::Continuous
	  (hooray for undocumented features :P)

0.06 Thu Jan 25 14:15:00 EST 2001 (UNRELEASED)
	- Skipped a version, oh well.
	- Add the sub-directories' test suites to MANIFEST (doh)
	- Fix the calling conventions for uttproc_rawdata() and
	  uttproc_result().
	- Add a testsuite for Speech::Recognizer::SPX itself.
	- Add demos of using Audio::SPX::Continuous with select().
	- Add proper path searching for Sphinx2 libraries in
	  Makefile.PL and use the exported $SPHINXDIR variable for
	  defaults.
	- Fix 'undefined value' warnings in Audio::SPX read functions.
	- Allow the use of hash references for fbs_init().

0.04_99 Wed Dec 13 17:02:59 EST 2000 (UNRELEASED)
	- Split off feature extraction and audio modules, link with
	  necessary libraries
	- Changed name to Speech::Recognizer::SPX to comply with the
	  Sphinx2 license and fit into the CPAN namespace better
	- Add exporter functionality to Speech::Recognizer::SPX, and
	  put together some nice tags of things to export.
	- The only thing missing before this can be released is
	  documentation and a test suite, I think.

0.04 Mon Dec  4 17:34:34 EST 2000 (UNRELEASED)
	- Add cepstral analysis (Sphinx2::FE) methods.
	- Make Sphinx2::Audio::init and Sphinx2::Audio::Continuous::init
	  class methods.
	- Fix tty-continuous.

0.03 Sun Dec 03 14:17:46 2000 (UNRELEASED)
	- Add a server module and a demo of streaming audio to Sphinx
	  over the network.  Remove the really broken and stupid
	  pipe-demo.pl :)

0.02_01  Fri Dec 01 11:29:46 2000 (UNRELEASED)
	- Add copyright notices.

0.02  Fri Dec 01 00:42:57 2000 (UNRELEASED)
        - Fix the interface, removing evil autoload.
	- Add the option to specify one's own ARGV to fbs_init()
	- Add a hacked-up continuous-audio-from-random-file API.
	- Add two demos.

0.01  Wed Nov 29 12:25:19 2000 (UNRELEASED)
	- original version; created by h2xs 1.19



test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl test.pl'

######################### We start with some black magic to print on failure.

# Change 1..1 below to 1..last_test_to_print .
# (It may become useful if the test is moved to ./t subdirectory.)

BEGIN { $| = 1; print "1..5\n"; }
END {print "not ok 1\n" unless $loaded;}
use Speech::Recognizer::SPX qw(:uttproc :fbs $SPHINXDIR);
use Speech::Recognizer::SPX::Server;
use Config;
$loaded = 1;
print "ok 1\n";

######################### End of black magic.

# Insert your test code below (better if it prints "ok 13"
# (correspondingly "not ok 13") depending on the success of chunk 13
# of the test code):

fbs_init({-live		=> 'TRUE',
	  -samp		=> 16000,
	  -adcin	=> 'TRUE',
	  -ctloffset	=> 0,
	  -ctlcount	=> 100000000,
	  -cepdir	=> "$SPHINXDIR/model/lm/turtle",
	  -datadir	=> "$SPHINXDIR/model/lm/turtle",
	  -agcmax	=> 'FALSE',
	  -langwt	=> 6.5,
	  -fwdflatlw	=> 8.5,
	  -rescorelw	=> 9.5,
	  -ugwt		=> 0.5,
	  -fillpen	=> 1e-10,
	  -silpen	=> 0.005,
	  -inspen	=> 0.65,
	  -top		=> 1,
	  -topsenfrm	=> 3,
	  -topsenthresh	=> -70000,
	  -beam		=> 2e-06,
	  -npbeam	=> 2e-06,
	  -lpbeam	=> 2e-05,
	  -lponlybeam	=> 0.0005,
	  -nwbeam	=> 0.0005,
	  -fwdflat	=> 'FALSE',
	  -fwdflatbeam	=> 1e-08,
	  -fwdflatnwbeam=> 0.0003,
	  -bestpath	=> 'TRUE',
	  -kbdumpdir	=> $SPHINXDIR,
	  -lmfn		=> "$SPHINXDIR/model/lm/turtle/turtle.lm",
	  -dictfn	=> "$SPHINXDIR/model/lm/turtle/turtle.dic",
	  -phnfn	=> "$SPHINXDIR/model/hmm/6k/phone",
	  -mapfn	=> "$SPHINXDIR/model/hmm/6k/map",
	  -hmmdir	=> "$SPHINXDIR/model/hmm/6k",
	  -hmmdirlist	=> "$SPHINXDIR/model/hmm/6k",
	  -ndictfn	=> "$SPHINXDIR/model/hmm/6k/noisedict",
	  '-8bsen'	=> 'TRUE',
	  -sendumpfn	=> "$SPHINXDIR/model/hmm/6k/sendump",
	  -cbdir	=> "$SPHINXDIR/model/hmm/6k"});


sub BIG_ENDIAN () { $Config{byteorder} eq '4321' }
sub swab16 { $_[0] =~ s/(.)(.)/$2$1/gs }

open WAV, "<testwav.raw" or die "couldn't open testwav.raw: $!";
uttproc_begin_utt();
my $count = 0;
while (defined(my $b = read WAV, my($buf), 4096)) {
    last if $b == 0;
    $count += $b;
    swab16($buf) if BIG_ENDIAN;
    uttproc_rawdata($buf, 1)
	or die "uttproc_rawdata failed";
    if ($count > 50000 and $count < 54096) {
	my ($fr, $hyp) = uttproc_partial_result();
	if ($hyp =~ /TO THE LAB/i) {
	    print "ok 2\n";
	} else {
	    print "not ok 2\n";
	}
	my ($fr, $hypseg) = uttproc_partial_result_seg();
	if ($hyp eq $hypseg->sent()) {
	    print "ok 3\n";
	}
	else {
	    print "not ok 3\n";
	}
    }
}
uttproc_end_utt();
my ($fr, $hyp) = uttproc_result(1);
print "hyp $hyp frames $fr\n";
my @nbest = search_get_alt(5);
my $hyp2 = $nbest[0]->sent();
print "nbest[0] $hyp2\n";
my $segs = $nbest[0]->segs();
foreach my $seg (@$segs) {
    printf "  Start frame %d end frame %d word %s\n",
	$seg->sf, $seg->ef, $seg->word;
}
fbs_end();

if ($hyp =~ /GO TO THE LAB SAY HELLO TO ROBOMAN QUIT/i) {
    print "ok 4\n";
} else {
    print "not ok 4\n";
}

if ($hyp eq $hyp2) {
    print "ok 5\n";
} else {
    print "not ok 5\n";
}


SPX.pm
use strict;

# Speech::Recognizer::SPX:  Perl interface to the Sphinx-II speech recognizer.

# Copyright (c) 2000 Cepstral LLC.
#
# This module is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself.
#
# Written by David Huggins-Daines <dhuggins@cs.cmu.edu>

package Speech::Recognizer::SPX;
use Carp;
use vars qw($VERSION @ISA @EXPORT_OK %EXPORT_TAGS $AUTOLOAD);

require Exporter;
require DynaLoader;
require Speech::Recognizer::SPX::Config;

@ISA = qw(Exporter DynaLoader);
@EXPORT_OK = qw(
		fbs_init
		fbs_end

		uttfile_open
		uttproc_begin_utt
		uttproc_rawdata
		uttproc_cepdata
		uttproc_end_utt
		uttproc_abort_utt
		uttproc_stop_utt
		uttproc_restart_utt
		uttproc_result
		uttproc_result_seg
		uttproc_partial_result
		uttproc_partial_result_seg
		uttproc_get_uttid
		uttproc_set_auto_uttid_prefix
		uttproc_set_lm
		uttproc_lmupdate
		uttproc_set_context
		uttproc_set_rawlogdir
		uttproc_set_mfclogdir
		uttproc_set_logfile
		search_get_alt

		lm_read
		lm_delete

		BG_SEG_SZ
		LM3G_ACCESS_BG
		LM3G_ACCESS_ERR
		LM3G_ACCESS_TG
		LM3G_ACCESS_UG
		LOG_BG_SEG_SZ
		NO_WORD
		NUM_COEFF
		NUM_SMOOTH
		OFFSET_LIKELIHOOD
		SPHINXP_PORT

		$SPHINXDIR
);
%EXPORT_TAGS = (
		fbs => [qw(
			   fbs_init
			   fbs_end
			  )
		       ],
		uttproc => [qw(
			       uttfile_open
			       uttproc_begin_utt
			       uttproc_rawdata
			       uttproc_cepdata
			       uttproc_end_utt
			       uttproc_abort_utt
			       uttproc_stop_utt
			       uttproc_restart_utt
			       uttproc_result
			       uttproc_result_seg
			       uttproc_partial_result
			       uttproc_partial_result_seg
			       uttproc_get_uttid
			       uttproc_set_auto_uttid_prefix
			       uttproc_set_lm
			       uttproc_lmupdate
			       uttproc_set_context
			       uttproc_set_rawlogdir
			       uttproc_set_mfclogdir
			       uttproc_set_logfile
			       search_get_alt
			      )
			   ],
		lm => [
		       qw(
			  lm_read
			  lm_delete
			 )
		      ],
	       );

$VERSION = 0.08_01;

sub AUTOLOAD {
    # This AUTOLOAD is used to 'autoload' constants from the constant()
    # XS function.  If a constant is not found then control is passed
    # to the AUTOLOAD in AutoLoader.

    my $constname;
    ($constname = $AUTOLOAD) =~ s/.*:://;
    croak "& not defined" if $constname eq 'constant';
    my $val = constant($constname, @_ ? $_[0] : 0);
    if ($! != 0) {
	if ($! =~ /Invalid/) {
	    croak "Undefined subroutine $AUTOLOAD";
	}
	else {
	    croak "Your vendor has not defined Speech::Recognizer::SPX macro $constname";
	}
    }
    no strict 'refs';
    if ($] >= 5.00561) {
	*$AUTOLOAD = sub () { $val };
    } else {
	*$AUTOLOAD = sub { $val };
    }
    goto &$AUTOLOAD;
}

bootstrap Speech::Recognizer::SPX $VERSION;

package Speech::Recognizer::SPX::Hypothesis;

# CAUTION: These constants must match the ordering in SPX.xs (new_hyp_sv)
use constant SENT => 0;
use constant UTTID => 1;
use constant SENSCALE => 2; # Sphinx3 will use this
use constant ASCR => 3;
use constant LSCR => 4;
use constant SEGS => 5;

sub sent { $_[0][SENT] }
sub uttid { $_[0][UTTID] }
sub senscale { $_[0][SENSCALE] }
sub ascr { $_[0][ASCR] }
sub lscr { $_[0][LSCR] }
sub segs { $_[0][SEGS] }

package Speech::Recognizer::SPX::Segment;

# CAUTION: These constants must match the ordering in SPX.xs (new_seg_sv)
use constant WORD => 0;
use constant SF => 1;
use constant EF => 2;
use constant ASCR => 3;
use constant LSCR => 4;
use constant FSG_STATE => 5;
use constant FSG_STATE_TO => 5;
use constant CONF => 6;
use constant LATDEN => 7;
use constant PHONE_PERP => 8;
use constant FSG_STATE_FROM => 9;

sub word { $_[0][WORD] }
sub sf { $_[0][SF] }
sub ef { $_[0][EF] }
sub ascr { $_[0][ASCR] }
sub lscr { $_[0][LSCR] }
sub fsg_state { $_[0][FSG_STATE] }
sub fsg_state_to { $_[0][FSG_STATE] }
sub conf { $_[0][CONF] }
sub latden { $_[0][LATDEN] }
sub phone_perp { $_[0][PHONE_PERP] }
sub fsg_state_from { $_[0][FSG_STATE_FROM] }

1;
__END__

=head1 NAME

Speech::Recognizer::SPX - Perl extension for the Sphinx2 speech recognizer

=head1 SYNOPSIS

  use Speech::Recognizer::SPX qw(:fbs :uttproc)
  fbs_init([arg1 => $val, arg2 => $val, ...]);
  uttproc_begin_utt();
  uttproc_end_utt();
  fbs_end();

=head1 DESCRIPTION

This module provides a Perl interface to the Sphinx-II speech
recognizer library.

Warning!  This interface is subject to change.  It's currently a bit
clunky because of the way the Sphinx-II library is structured, and
that will probably change (for the better, I hope) over time.

When the interface changes, future versions of this documentation will
point out how it has changed and how to deal with this.

=head1 USING THIS MODULE

  use Speech::Recognizer::SPX qw(:fbs :uttproc :lm);

Because most parts of the Sphinx-II library contain a lot of global
internal state, it makes no sense to use an object-oriented interface
at this time.  However I don't want to clobber your namespace with a
billion functions you may or may not use.  To make things easier on
your typing hands, the available functions have been grouped in to
tags representing modules inside the library itself.  These tags and
the functions they import are listed below.

=over 4

=item C<:fbs>

This is somewhat of a misnomer - FBS stands for Fast Beam Search, but
in actual fact this module (the fbs_main.c file in Sphinx-II) just
wraps around the other modules in sphinx (one of which actually does
fast beam search :-) and initializes the recognizer for you.
Functions imported by this tag are:

  fbs_init
  fbs_end

=item C<:uttproc>

This is the utterance processing module.  You feed it data (either raw
audio data or feature data - which currently means vectors of
mel-frequency cepstral coefficients), and it feeds back search
hypotheses based on a language model.  Functions imported by this tag
are:

  uttfile_open
  uttproc_begin_utt
  uttproc_rawdata
  uttproc_cepdata
  uttproc_end_utt
  uttproc_abort_utt
  uttproc_stop_utt
  uttproc_restart_utt
  uttproc_result
  uttproc_result_seg
  uttproc_partial_result
  uttproc_partial_result_seg
  uttproc_get_uttid
  uttproc_set_auto_uttid_prefix
  uttproc_set_lm
  uttproc_lmupdate
  uttproc_set_context
  uttproc_set_rawlogdir
  uttproc_set_mfclogdir
  uttproc_set_logfile
  search_get_alt

=item C<:lm>

This is the language model module.  It loads and unloads language models.

  lm_read
  lm_delete

=back

=head1 STARTUP

  fbs_init(\@args);

The C<fbs_init> function is the main entry point to the Sphinx
library.  If given no arguments, it will snarf options from the global
C<@ARGV> array (because that's what its C equivalent does).  To make
life easier, and to entice people to write Sphinx programs in Perl
instead of C, we also give you a way around this by allowing you to
also pass a reference to an array whose contents are arranged in the
same way C<@ARGV> might be, i.e. a list of option/value pairs.

To make things pretty, you can use the magical => operator, like this:

  fbs_init([samp => 16000,
            datadir => '/foo/bar/baz']);

Note that you can omit the leading dash from argument names (if you
like).

Calling this function will block your process for a long time and
print unbelievable amounts of debugging gunk to STDOUT and STDERR.
This will get better eventually.

This function has a large number of options.  Someday they will be
documented.  Until then, either look in the example code, or go
straight to the source, namely the C<param> variable in
C<src/libsphinx2/fbs_main.c> and the C<kb_param> variable in
C<src/libsphinx2/kb_main.c>.

=head1 FEED ME

  uttproc_begin_utt() or die;
  uttproc_rawdata($buf [, $block]) or die;
  uttproc_cepdata(\@cepvecs [, $block]) or die;
  uttproc_end_utt() or die;

To actually recognize some speech data, you use the functions exported
by the C<:uttproc> tag.  Before calling any of them, you must
successfully call C<uttproc_begin_utt>, or Bad Things are certain to
happen (I can't speculate on exactly what things, but I'm sure they're
bad).

You should call C<uttproc_begin_utt> before each distinct utterance
(to the extent that you can predict when individual utterances begin
or end, of course...), and C<uttproc_end_utt> at the end of each.

After calling C<uttproc_begin_utt>, you can pass either raw audio data
or cepstral feature vectors (see L<Audio::MFCC>), using
C<uttproc_rawdata> or C<uttproc_cepdata>, respectively.  Due to the
way feature extraction works, you cannot mix the two types of data
within the same utterance.

If live mode is in effect (i.e. C<-livemode => TRUE> was passed to
C<fbs_init>), the optional C<$block> parameter controls whether these
functions will return immediately after processing a single frame of
data, or whether they will process all pending frames of data.  If you
need partial results, you probably want to pass a non-zero value
(FIXME: should be a true value but I don't know how to test for truth
in XS code) for C<$block>, though this may increase latency elsewhere
in the system.

Unfortunately, it appears that there is no specific function to flush
all unprocessed frames before getting a partial result.  Calling
uttproc_rawdata with an empty C<$buf> and C<$block> non-zero seems to
have the desired effect.

=head1 GETTING RESULTS

  my ($frames, $hypothesis) = uttproc_result($block);
  my ($frames, $hypothesis) = uttproc_partial_result();
  my ($frames, $hypseg) = uttproc_result_seg($block);
  my ($frames, $hypseg) = uttproc_partial_result_seg();
  my $hypothesis = $hypseg->sent;
  my $segs = $hypseg->segs;
  my @nbest = search_get_alt($n); # Must call uttproc_result first!
  foreach my $nhyp (@nbest) {
    my $nsent = $nhyp->sent;
    my $nsegs = $nhyp->segs;
    print "Hypothesis: $nsent\n";
    foreach my $seg (@$nsegs) {
      printf "  Start frame %d end frame %d word %s\n",
             $seg->sf, $seg->ef, $seg->word;
    }
  }

At any point during utterance processing, you may call
C<uttproc_partial_result> to obtain the current "best guess".  Note
that this function does B<not> flush unprocessed frames, so you might
want to use the trick mentioned above to do so before calling it if
you are operating in non-blocking mode.

By contrast, you may not call C<uttproc_result> until after you have
called C<uttproc_end_utt> (or C<uttproc_abort_utt> or also possibly
C<uttproc_stop_utt>).  The C<$block> flag is also optional here, but I
strongly suggest you use it.

The functions C<uttproc_result_seg> and C<uttproc_partial_result_seg>
functions work similarly except that instead of returning a string,
they return a C<Speech::Recognizer::SPX::Hypothesis> object which
contains probability and word segmentation information.  You can
access its fields with the following accessor functions:

  sent
  senscale
  ascr
  lscr
  segs

The C<sent> field contains the string representation of the
hypothesis, and is equivalent to the string returned by
C<uttproc_result>.  The C<senscale>, C<ascr>, and C<lscr> fields are
currently unimplemented.

The C<segs> field contains a reference to an array of
C<Speech::Recognizer::SPX::Segment> objects.  Each of these objects
contains fields which can be accessed with the following accessors:

  word
  sf
  ef
  ascr
  lscr
  conf
  latden
  phone_perp
  fsg_state_to
  fsg_state_from

The C<word> field contains the string representation of the word.  The
C<sf> and C<ef> fields contain the start and end frames for this word.
The C<ascr> and C<lscr> fields contain the acoustic and language model
scores for the word.  The C<fsg_state_from> and C<fsg_state_to> fields
indicate the finite-state grammar states in which this entry starts
and terminates, if a finite-state grammar is used.  The C<latden>
field contains the average lattice density for this word, while the
C<phone_perp> contains the average phoneme perplexity.  The C<conf>
field contains a confidence score which is an estimated probability
that this word was recognized correctly.

You can also obtain an N-best list of hypotheses using the
C<search_get_alt> function.  This function returns a list of the
number of hypotheses requested, or as many as can be found.  Each
element in this list is a C<Speech::Recognizer::SPX::Hypothesis>
object like the above, except that the acoustic/language model score
is not filled in.

=head1 FIDDLY BITS

Changing language models, etc, etc...  This documentation is under
construction.

=head1 EXAMPLES

For now there are just some example programs in the distribution.

=head1 AUTHOR

David Huggins-Daines <dhuggins@cs.cmu.edu>.  Support for N-best
hypotheses was funded by SingleTouch Interactive, Inc
(http://www.singletouch.net/).

=head1 SEE ALSO

perl(1), L<Speech::Recognizer::SPX::Server>, L<Audio::SPX>, L<Audio::MFCC>

=cut


SPX/Server.pm
# -*- cperl -*-
use strict;

# Speech::Recognizer::SPX::Server: Perl module for writing Sphinx-II
# streaming audio servers.

# Copyright (c) 2000 Cepstral LLC.
#
# This module is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself.
#
# Written by David Huggins-Daines <dhuggins@cs.cmu.edu>

package Speech::Recognizer::SPX::Server;
use Speech::Recognizer::SPX qw(:fbs :uttproc $SPHINXDIR);
use Audio::SPX;
use Time::HiRes qw(usleep);
use Fcntl;
use Errno;

use vars qw($VERSION);
$VERSION=0.02;

my %defaults = (-live		=> 'TRUE',
		-samp		=> 16000,
		-adcin		=> 'TRUE',
		-ctloffset	=> 0,
		-ctlcount	=> 100000000,
		-cepdir		=> "$SPHINXDIR/model/lm/turtle",
		-datadir	=> "$SPHINXDIR/model/lm/turtle",
		-agcmax		=> 'FALSE',
		-langwt		=> 6.5,
		-fwdflatlw	=> 8.5,
		-rescorelw	=> 9.5,
		-ugwt		=> 0.5,
		-fillpen	=> 1e-10,
		-silpen		=> 0.005,
		-inspen		=> 0.65,
		-top		=> 1,
		-topsenfrm	=> 3,
		-topsenthresh	=> -70000,
		-beam		=> 2e-06,
		-npbeam		=> 2e-06,
		-lpbeam		=> 2e-05,
		-lponlybeam	=> 0.0005,
		-nwbeam		=> 0.0005,
		-fwdflat	=> 'FALSE',
		-fwdflatbeam	=> 1e-08,
		-fwdflatnwbeam	=> 0.0003,
		-bestpath	=> 'TRUE',
		-kbdumpdir	=> $SPHINXDIR,
		-lmfn		=> "$SPHINXDIR/model/lm/turtle/turtle.lm",
		-dictfn		=> "$SPHINXDIR/model/lm/turtle/turtle.dic",
		-phnfn		=> "$SPHINXDIR/model/hmm/6k/phone",
		-mapfn		=> "$SPHINXDIR/model/hmm/6k/map",
		-hmmdir		=> "$SPHINXDIR/model/hmm/6k",
		-hmmdirlist	=> "$SPHINXDIR/model/hmm/6k",
		-ndictfn	=> "$SPHINXDIR/model/hmm/6k/noisedict",
		'-8bsen'	=> 'TRUE',
		-sendumpfn	=> "$SPHINXDIR/model/hmm/6k/sendump",
		-cbdir		=> "$SPHINXDIR/model/hmm/6k");

sub init {
    my ($this, $args, $sock, $log, $verbose) = @_;
    my $class = ref $this || $this;

    unless (defined $log) {
	local *LOG;
	open LOG, ">&STDERR" or die "can't dup: $!";
	$log = *LOG;
	# Readjust die() so we actually see messages ;(
	$SIG{__DIE__} = sub { print $log @_; exit $! || ($? >> 8) || 255 };
    }

    print $log "initializing sphinx2\n" if $log;
    my @argv = (%defaults, %$args, verbose => $verbose);
    {
	my $i = 0;
	foreach (@argv) {
	    unless ($i++ % 2) {
		# Make argument names look like arguments to sphinx
		$_ = "-$_" unless $_ =~ /^-/;
	    }
	}
    }
    fbs_init(\@argv)
	or return undef;

    my $sps = $args->{-samp} || $defaults{-samp};
    my $self = { sock => $sock, log => $log,
		 sps => $sps, sockflags => 0,
		 timeout => 1000,
		 verbose => $verbose };

    bless $self, $class;
}

sub sock { return $_[1] ? ($_[0]->{sock} = $_[1]) : $_[0]->{sock} }
sub logfh { return $_[1] ? ($_[0]->{log} = $_[1]) : $_[0]->{log} }
sub timeout { return $_[1] ? ($_[0]->{timeout} = $_[1]) : $_[0]->{timeout} }

sub calibrate {
    my ($self) = @_;
    my ($sock, $sps, $log, $verbose) =
	@{$self}{qw/sock sps log verbose/};

    print $log "calibrating\n" if $log;

    # Use blocking mode for calibration
    print $log "initializing continuous audio\n" if $log;
    my $res = undef;
    my $cad = Audio::SPX::Continuous->init_nbfh($sock, $sps);
    if (defined($cad)) {
	$res = $cad->calib;
	$self->{cad} = $cad;
    }

    return $res;
}

sub next_utterance {
    my ($self, $cb_listen, $cb_not_listen, $audio_fh) = @_;

    my ($cad, $sock, $sps, $log, $verbose) =
	@{$self}{qw/cad sock sps log verbose/};

    my $adbuf = "";

    # Use non-blocking mode for reading data (FIXME: should use select really...)
    fcntl $sock, F_GETFL, $self->{sockflags};
    fcntl $sock, F_SETFL, $self->{sockflags} | O_NONBLOCK;

    print $log "waiting for audio\n" if $log;
    my $s;
    while (defined($s = $cad->read($adbuf, 2048)) && $s == 0) {
	usleep 50_000;
    }
    goto failure unless defined $s;

    my $ts = $cad->read_ts;
    print $log "listening at $ts\n" if $log;
    $cb_listen->($ts) if defined $cb_listen;

    print $audio_fh $adbuf if defined $audio_fh;

    uttproc_begin_utt() or goto failure;
    uttproc_rawdata($adbuf, 0) or goto failure;

    while (1) {
	$adbuf = "";
	$s = $cad->read($adbuf, 2048) or goto failure;
	if ($s == 0) {
	    last if $cad->read_ts - $ts > int($sps * $self->{timeout} / 1000);
	    usleep(20_000);
	} else {
	    $ts = $cad->read_ts;
	    print $audio_fh $adbuf if defined $audio_fh;
	    my $rem = uttproc_rawdata($adbuf, 0)
		or goto failure;
	}
    }

    $cad->reset;
    fcntl $sock, F_SETFL, $self->{sockflags};

    print $log "done listening at $ts\n" if $log;
    $cb_not_listen->($ts) if defined $cb_not_listen;

    uttproc_end_utt() or goto failure;

    goto failure
	unless (my ($fr, $hyp) = uttproc_result(1));

    print $log "text is $hyp\n" if $log;
    return $hyp;

 failure:
    # Need to restore this for the caller
    fcntl $sock, F_SETFL, $self->{sockflags};
    return undef;
}

sub fini {
    fbs_end();
}

1;
__END__

=head1 NAME

Speech::Recognizer::SPX::Server - Perl module for writing streaming audio speech recognition servers using Sphinx2

=head1 SYNOPSIS

  my $sock = new IO::Socket(... blah blah blah ...);
  my $log = new IO::File('server.log');
  my $audio_fh = new IO::File('speech.raw');
  my $srvr
      = Speech::Recognizer::SPX::Server->init({ -arg => val, ... }, $sock, $log, $verbose)
        or die "couldn't initialize sphinx2: $!";

  my $client = new IO::Socket;
  while (accept $sock, $client) {
      next unless fork;
      $srvr->sock($client);
      $srvr->calibrate or die "couldn't calibrate audio stream: $!";
      while (!$done && defined(my $txt
			= $srvr->next_utterance(sub { print $log "listening\n" },
						sub { print $log "not listening\n },
						$audio_fh))) {
	  print "recognized text is $txt\n";
	  ...
      }
      $srvr->fini or die "couldn't shut down server: $!";
      exit 0;
  }

=head1 DESCRIPTION

This module encapsulates a bunch of the stuff needed to write a
Sphinx2 server which takes streaming audio as input on an arbitrary
filehandle.  It's not meant to be flexible or transparent - if you
want that, then read the code and write your own server program using
just the Speech::Recognizer::SPX module.

The interface is vaguely object-oriented, but unfortunately it is
presently not possible to create multiple instances of
Speech::Recognizer::SPX::Server within the same process, due to severe
limitations of the underlying Sphinx-II library.  You can, however,
create multiple distinct servers with judicious use of C<fork>, as
shown in the example above.

It is possible that this will be fixed in a future release of Sphinx-II.

=head1 METHODS

=over 4

=item C<init>

  my $srvr = Speech::Recognizer::SPX::Server->init(\%args, $sock, $log, $verbose);

C<%args> is a reference to a hash of argument => value pairs, exactly
like the arguments you would pass on the command line to one of the
sphinx example programs.  Argument names can be given either with or
without a leading dash.

C<$sock> is a socket or other filehandle (could be anything, really)
on which the server will read audio data.  This argument is optional
and not needed to initialize the server - you can set it later with
the C<sock> accessor.

C<$log> is a filehandle on which the server module will log messages.
This argument is optional.  Without a filehandle to log on, these
messages (boring things like "started listening at $foo") will not be
printed.

C<$verbose> determines the verbosity level of the Sphinx library.
Currently, due to limitations in the Sphinx-II library, there are only
two options for this value, namely a true value for 'be insanely
verbose', or a false value for 'say nothing at all'.

=item C<calibrate>

  $srvr->calibrate;

Calibrates the noise threshold for the continuous audio stream
(i.e. figures out when it should listen and when it shouldn't).  This
requires you to actually have a ready and willing source of input on
the socket you set in C<init> or with C<sock>.

=item C<next_utterance>

  my $text = $srvr->next_utterance($cb_listen, $cb_not_listen, $audio_fh);

Waits for and recognizes the next utterance in the data stream.  All
arguments are optional:

C<$cb_listen> is a reference to (or name of, but I encourage you not
to do that) a subroutine to be called when the recognizer has detected
speech input.

C<$cb_not_listen> is a reference to (or name of) a subroutine to be
called when the recognizer has detected the end of speech input.

Obviously this presumes a request/response model for your application.
If you need to be able to get partial results then you'll have to wait
for me to support them (which will undoubtedly happen sooner or
later), or write your own module.  Sorry.

C<$audio_fh> is a filehandle to which to save the speech data - this
may come in handy for debugging, or if you would like to only record
the user talking and not the hours and hours of silence in between.

=item C<fini>

Shuts down the Sphinx-II recognizer.  Doesn't close the socket or
anything though, you have to do that yourself.

=back

=head1 ACCESSORS

=over 4

=item C<sock>

  my $sockfh = $srvr->sock;
  $srvr->sock(\*FOO);

Sets or gets the socket on which the server reads audio data.

=item C<logfh>

  my $logfh = $srvr->logfh;
  $srvr->log(\*BAR);

Sets or gets the filehandle on which the server logs messages (if it's
being verbose).

=item C<timeout>

  $srvr->timeout(200);

Sets/gets the amount of time (in milliseconds) to wait after the end
of speech-level input before processing an utterance.  Default is one
second.

=back

=head1 SEE ALSO

perl(1), L<Speech::Recognizer::SPX>.

=head1 AUTHOR

David Huggins-Daines <dhuggins@cs.cmu.edu>

=cut
